### ISSUE TITLE
WebGPURenderer: Increase performance

### ISSUE THREAD

**Issue by sunag** [mrdoob/three.js#26673]
### Description

I started a work to increase the performance in `WebGPURenderer` and it will work for both `Backends`, it's about a better management of `Bindings` and `Cache`.

### Solution

- [ ] Check Material Cache Key only after the version difference -- `material.needsUpdate=true`
- [ ] Reuse `NodeBuilder` based from `Cache Key`, currently the cache uses Material/Geometry as reference. For example, this means that two Materials of the same type with only a different color will be compiled twice instead of reuse the `NodeBuilder`, this sharing already happens for the GPU Program, but as the code is dynamically generated this optimization process should be API oriented for the Nodes.
- [ ] Separate bindings groups for `FRAME`, `RENDER`, `OBJECT` groups. Node System already offers this support, we would have to take it to the level of optimization for the Bindings.
- [ ] Node System allows Auto-Instancing with a certain flexibility, this has already been adopted in other engines like Unreal and we may be implementing it soon as well. -- https://github.com/mrdoob/three.js/pull/26640

### Alternatives

--

**Comment by danrossi** [comment]
I was about to comment. I just made new builds to figure out the iife bundle issue. And super low resources with video textures.

https://danrossi.github.io/three-webgpu-renderer/tests/webgpu_video_panorama_equirectangular.html

and using video frame callback

https://danrossi.github.io/three-webgpu-renderer/tests/webgpu_video_panorama_equirectangular_30fps.html

**Comment by PhilCrowther** [comment]
There is apparently still an issue with WebGPUrenderer and shadows.  On a program with a single DirectionalLight  that has shadow definitions, and that uses non-emissive materials (e.g. Lambert and Phong), switching to WebGPUrenderer will cause a massive slowdown.  Oddly, if I remove the light, WebGPUrender will cause all the materials to become emissive - with the result that shadows are no longer an issue and the program operates at normal speed.  (This has nothing to do with the use of NodeMaterials as I first thought since the same thing happens with regular Materials.)  If it helps, I also had to add a bias definition to fix the appearance of some materials.

**Comment by sunag** [comment]
@PhilCrowther Could it be related to [this](https://github.com/mrdoob/three.js/pull/27068) ?

**Comment by PhilCrowther** [comment]
@sunag Not sure.  I have created a [small demo on CodePen](https://codepen.io/Phil_Crowther/pen/rNPexYz).
To show the difference in performance and results, disable the DirectionalLight (line 82: scene.add(DirLight);).
But, let me know if this is just a matter of adding a few lines to my program.

**Comment by sunag** [comment]
Hmm... thanks for the example. It looks like there is a conflict in `frameId`.
Try to remove any `requestAnimationFrame(rendAll);` and add `renderer.setAnimationLoop(rendAll);`

Related: https://github.com/mrdoob/three.js/pull/24755

Anyway, I'll be checking for a fix for this.

**Comment by PhilCrowther** [comment]
Yes, that appears to have fixed the problems I was having with various versions of the program.  I had to disable a smoke emitter, but I plan to update that anyways using the WebGPU Particles example.  Thanks for the quick response!

**Comment by greggman** [comment]
I don't know if any of this is useful to three.js but [this article has some ideas for WebGPU performance](https://webgpufundamentals.org/webgpu/lessons/webgpu-optimization.html).

There are also some hopefully equivalent WebGL examples linked at the bottom of [this page](https://webgpufundamentals.org/webgpu/lessons/webgpu-from-webgl.html) for comparison

Certainly, WebGPU implementaitions should find ways to increase perf more but at a glance it seems like it should be possible to go as fast or faster then WebGL.

**Comment by sunag** [comment]
I spent a few days analyzing the current performance stage, here are some observations:

We still have a heavy CPU usage load to reduce, which means the new features we added haven’t kept pace with the cache system.

The system for detecting uniforms and updates needs to be more efficient than the current check. Currently the values are tested during rendering, trigger an event for an update flag, which would be much more efficient. This may be aligned with this update https://github.com/mrdoob/three.js/pull/28328

I also noticed that we need to consolidate equivalent materials/geometry and make RenderObject more abstract with respect to the object, sharing material/geometry across objects with a compatible configuration. In this sense, a RenderObject could have several objects, making the rendering of similar objects more effective and avoiding “duplicate” commands to the GPU, as in the case of using thousands of repeated objects without instancing. This should address https://github.com/mrdoob/three.js/issues/30560

/cc @Mugen87

**Comment by Mugen87** [comment]
Yes, I think we have reached a point where we should focus on making the existing features faster, consolidate/refactor internals and maybe some part of the public API. 

Now that more and more devs are starting to use the new renderer and material system, more errors or API inconsistencies are reported. We should focus on them as well.

I actually wanted to start with a basic node class for SSGI/SSRT to enhance our post processing but I delay that now to focus on the open issues^^.

> This should address https://github.com/mrdoob/three.js/issues/30560

For fixing this particular issue, I believe there is no way around in how per-objects UBOs are managed. I wanted to experiment with the WebGL backend with a single buffer that holds all per-object uniform data per render calls so you just end up a single `bufferSubData()` call. We then switch to `bindBufferRange()` to create a view into this buffer for each UBO. Not sure if this works and how it performs but it's maybe worth giving it a shot.

**Comment by sunag** [comment]
You can try isolated a group through `uniform().setGroup()`, to obtain a reduced buffer according to the uniform, this should have a similar result to a sub-data. I did this for test with some nodes and got a better result, but still the CPU load is still large enough between one draw and another, so condensing several similar meshes into a single RenderObject should be the solution to reducing these node update calls.

But in fact, all optimization will help in this path.

**Comment by Mugen87** [comment]
An AI analyzed performance trace has been shared in https://github.com/mrdoob/three.js/issues/32916#issue-3880675765.
